cmd/set-nas-status/main.go

```diff
@@ -22,191 +22,92 @@ import (
 	"os"
 	"strings"
 
-	"github.com/spf13/cobra"
-	"github.com/spf13/pflag"
-	"github.com/spf13/viper"
+	"github.com/urfave/cli/v2"
 
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	coreclientset "k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/client-go/util/retry"
-	cliflag "k8s.io/component-base/cli/flag"
-	"k8s.io/component-base/term"
-	"k8s.io/klog/v2"
 
 	nascrd "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1"
 	nasclient "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1/client"
-	exampleclientset "github.com/kubernetes-sigs/dra-example-driver/pkg/example.com/resource/clientset/versioned"
+	"github.com/kubernetes-sigs/dra-example-driver/pkg/kubeclient"
+	"github.com/kubernetes-sigs/dra-example-driver/pkg/logging"
+	"github.com/kubernetes-sigs/dra-example-driver/pkg/nodeallocationstate"
 )
 
-type Flags struct {
-	kubeconfig *string
-	status     *string
-}
-
-type Config struct {
-	flags         *Flags
-	nascrd        *nascrd.NodeAllocationState
-	exampleclient exampleclientset.Interface
-}
-
 func main() {
-	command := NewCommand()
-	err := command.Execute()
-	if err != nil {
-		fmt.Printf("Error: %v\n", err)
+	if err := newApp().Run(os.Args); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
 	}
 }
 
-// NewCommand creates a *cobra.Command object with default parameters.
-func NewCommand() *cobra.Command {
-	cmd := &cobra.Command{
-		Use:  "set-nas-status",
-		Long: "set-nas-status sets the status of the NodeAllocationState CRD managed by the DRA driver for GPUs.",
-	}
-
-	flags := AddFlags(cmd)
-
-	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
-		// Bind an environment variable to each input flag
-		v := viper.New()
-		v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
-		v.AutomaticEnv()
-		cmd.Flags().VisitAll(func(f *pflag.Flag) {
-			if !f.Changed && v.IsSet(f.Name) {
-				val := v.Get(f.Name)
-				if err := cmd.Flags().Set(f.Name, fmt.Sprintf("%v", val)); err != nil {
-					klog.Errorf("Unable to bind environment variable to input flag: %v=%v", f.Name, val)
+func newApp() *cli.App {
+	var (
+		status  string
+		crFlags nodeallocationstate.Flags
+	)
+	var kubeClientFlags kubeclient.Flags
```

Comment:
> Why are two of these grouped and the third one not? 

An  oversight.

With this approach, how does the app get access to the values of the kubeclient and nas flags?

Reply:
> With this approach, how does the app get access to the values of the kubeclient and nas flags?

I am biased by how we do this in all of our NVIDIA components, but the following is how I would probably structure this  given your most recent changes (without a `newApp()` function at all).

```
type Flags struct {
	flags.KubeClientConfig
	flags.LoggingConfig
	flags.NodeAllocationStateConfig
	status string
}

func main() {
	flags := Flags{}

	app := cli.NewApp()
	app.Name = "set-nas-status",
	app.Usage = "set-nas-status sets the status of the NodeAllocationState CRD managed by the DRA driver for GPUs.",
	app.ArgsUsage = " "
	app.HideHelpCommand = true

	app.Flags = []cli.Flag{
		&cli.StringFlag{
			Name:     "status",
			Usage:    "The status to set [Ready | NotReady].",
			Required: true,
			Action: func(_ *cli.Context, value string) error {
				switch strings.ToLower(value) {
				case strings.ToLower(nascrd.NodeAllocationStateStatusReady):
					flags.status = nascrd.NodeAllocationStateStatusReady
				case strings.ToLower(nascrd.NodeAllocationStateStatusNotReady):
					flags.status = nascrd.NodeAllocationStateStatusNotReady
				default:
					return fmt.Errorf("unknown status: %s", value)
				}
				return nil
			},
			EnvVars: []string{"STATUS"},
		},
	}
	app.Flags = append( app.Flags, flags.KubeClientConfig.Flags()...)
	app.Flags = append( app.Flags, flags.LoggingConfig.Flags()...)
	app.Flags = append( app.Flags, flags.NodeAllocationStateConfig.Flags()...)

	app.Before = func(*cli.Context) error {
		// Call out to helper function rather than inline here
	}

	app.Action =  func(*cli.Context) error {
		// Call out to helper function rather than inline here
	}

	err := app.Run(os.Args)
	if err != nil {
		log.Fatalf(util.Capitalize(err.Error()))
	}
}
```

The main thing I want to make sure we maintain is a common structure across all three of the commands, and I think this would make this possible.

Reply:
There's some value in having a `newApp` function: theoretically, it enables writing unit tests. We don't do that now, but with the code in `main` we'll never be able to add them either. Was that perhaps the reason why it was there to begin with?

I don't mind moving everything into `main`, I just wanted to ask first.

Reply:
I'm fine with having a wrapper after initializing the app and before calling Run(). Anywhere before that seems unnecessary though.